src/reconstruction/pcm_cuda.cu

```diff
@@ -269,224 +270,179 @@ __global__ void PCM_Reconstruction_3D(Real *dev_conserved,
 
   // declare conserved variables for each stencil
   // these will be placed into registers for each thread
-  Real d, mx, my, mz, E;
-  #ifdef DE
-  Real ge;
-  #endif
   #ifdef SCALAR
   Real scalar[NSCALARS];
-  #endif
-
+  #endif  //SCALAR
 
-  int n_cells = nx*ny*nz;
+  int const n_cells = nx*ny*nz;
 
   // get a thread ID
-  int tid = threadIdx.x + blockIdx.x * blockDim.x;
-  int zid = tid / (nx*ny);
-  int yid = (tid - zid*nx*ny) / nx;
-  int xid = tid - zid*nx*ny - yid*nx;
-  int id = xid + yid*nx + zid*nx*ny;
+  int id = threadIdx.x + blockIdx.x * blockDim.x;
+  int xid, yid, zid;
+  cuda_utilities::compute3DIndices(id, nx, ny, xid, yid, zid);
 
-  // x direction
-  if (xid < nx-1 && yid < ny && zid < nz)
+  // Guard to avoid out of bounds threads
+  if (xid < nx && yid < ny && zid < nz)
   {
-    // retrieve appropriate conserved variables
-    id = xid + yid*nx + zid*nx*ny;
-    d  = dev_conserved[            id];
-    mx = dev_conserved[  n_cells + id];
-    my = dev_conserved[2*n_cells + id];
-    mz = dev_conserved[3*n_cells + id];
-    E  = dev_conserved[4*n_cells + id];
+    // ========================================
+    // Retrieve appropriate conserved variables
+    // ========================================
+    Real const d  = dev_conserved[            id];
+    Real const mx = dev_conserved[  n_cells + id];
+    Real const my = dev_conserved[2*n_cells + id];
+    Real const mz = dev_conserved[3*n_cells + id];
+    Real const E  = dev_conserved[4*n_cells + id];
     #ifdef SCALAR
-    for (int i=0; i<NSCALARS; i++) {
-      scalar[i] = dev_conserved[(5+i)*n_cells + id];
-    }
-    #endif
+      for (int i=0; i<NSCALARS; i++)
+      {
+        scalar[i] = dev_conserved[(5+i)*n_cells + id];
+      }
+    #endif  //SCALAR
+    #ifdef  MHD
+      Real cellCenteredBx, cellCenteredBy, cellCenteredBz;
```

Comment:
Because those are the cell centered fields and they're computed with the `mhd::utils::cellCenteredMagneticFields` function. That function was written before we had C++17 support and so it doesn't use structured binding so the centered magnetic fields had to be passed it and set. I'm going to try it with a structured binding return instead